ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullFilterNode.java

```diff
@@ -0,0 +1,823 @@
+  private boolean isKeyedQuery = false;
+  private Optional<WindowBounds> windowBounds;
+  private List<GenericKey> keyValues;
+  private Set<UnqualifiedColumnReferenceExp> keyColumns;
+  private Set<UnqualifiedColumnReferenceExp> systemColumns;
+  private List<GenericKey> inKeys;
+  private BitSet seenKeys;
+  private Object[] keyContents;
+  private boolean containsINkeys;
+
+  public PullFilterNode(
+      final PlanNodeId id,
+      final PlanNode source,
+      final Expression predicate,
+      final MetaStore metaStore,
+      final KsqlConfig ksqlConfig,
+      final boolean isWindowed
+  ) {
+    super(id, source.getNodeOutputType(), source.getSourceName(), source);
+
+    Objects.requireNonNull(predicate, "predicate");
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "ksqlConfig");
+    this.rewrittenPredicate = new StatementRewriteForMagicPseudoTimestamp().rewrite(predicate);
+    this.isWindowed = isWindowed;
+
+    // Basic validation of WHERE clause
+    validateWhereClause();
+
+    // Validation and extractions of window bounds
+    windowBounds = isWindowed ? Optional.of(extractWindowBounds()) : Optional.empty();
+
+    // Extraction of key and system columns
+    extractKeysAndSystemCols();
+
+    // Extraction of key values
+    keyValues = extractKeyValues();
+
+    // Compiling expression into byte code
+    this.addAdditionalColumnsToIntermediateSchema = shouldAddAdditionalColumnsInSchema();
+    this.intermediateSchema = PullLogicalPlanUtil.buildIntermediateSchema(
+        source.getSchema(), addAdditionalColumnsToIntermediateSchema, isWindowed);
+    compiledWhereClause = CodeGenRunner.compileExpression(
+        rewrittenPredicate,
+        "Predicate",
+        intermediateSchema,
+        ksqlConfig,
+        metaStore
+    );
+
+  }
+
+  public Expression getRewrittenPredicate() {
+    return rewrittenPredicate;
+  }
+
+  @Override
+  public LogicalSchema getSchema() {
+    return getSource().getSchema();
+  }
+
+  @Override
+  public SchemaKStream<?> buildStream(final KsqlQueryBuilder builder) {
+    throw new UnsupportedOperationException();
+  }
+
+  public ExpressionMetadata getCompiledWhereClause() {
+    return compiledWhereClause;
+  }
+
+  public boolean isKeyedQuery() {
+    return isKeyedQuery;
+  }
+
+  public boolean isWindowed() {
+    return isWindowed;
+  }
+
+  public List<GenericKey> getKeyValues() {
+    return keyValues;
+  }
+
+  public Set<UnqualifiedColumnReferenceExp> getKeyColumns() {
+    return keyColumns;
+  }
+
+  public Optional<WindowBounds> getWindowBounds() {
+    return windowBounds;
+  }
+
+  public boolean getAddAdditionalColumnsToIntermediateSchema() {
+    return addAdditionalColumnsToIntermediateSchema;
+  }
+
+  public LogicalSchema getIntermediateSchema() {
+    return intermediateSchema;
+  }
+
+  private void validateWhereClause() {
+    seenKeys = new BitSet(schema.key().size());
```

Comment:
I find the reuse, multiple times, of the field `seenKeys` to be confusing.  Why not just make this a field of `Validator` and add a getter for it to be checked below?